### PR TITLE
Draft: Feature/aip 6942 Async Select 

### PR DESCRIPTION
**Pull request by reno1979**
Created an Async Select based on React Select.

Based on: https://www.figma.com/file/ORa7CBIooPgZj6HsEPBxNR/Design-bibliotheek-(gepubliceerd)?node-id=1303%3A6122

- todo write test

Skipped the options with icons and the styling of multi-select

### CODE REVIEW

File: src/form/AsyncSelect/AsyncSelect.tsx [hunk 1]
```diff
@@ -0,0 +1,43 @@
...
+	zIndexMenu?: number;
+} & ComponentProps<typeof ReactSelectAsyncSelect>;
+
+// Todo: make this a reusable component, the typeahead above should be moved to AIP
+// Get possible streetnames by location https://api.data.amsterdam.nl/geosearch/bag/?lat=52.3912316&lon=4.954566&radius=100
+// Todo: what if API is down? -> show error message?
+// Todo: what if API is slow? -> show loading indicator?
+// Todo: what if API returns no results? -> show no results message?
+// Todo: what if API returns an error? -> show error message?
+// Todo: what if prefilled value is not available in API? -> show error message?
```
Comment (dlnr) [inline src/form/AsyncSelect/AsyncSelect.tsx:18]: Thats a bunch of todo's. So it doesn't do the API fetch yet?

Also perhaps we should add that it needs a react-hook-form controller to register this. 

I'm working with Mike on a spinner for the GPS component, we should look at adding loading state to the form? or Component ?

**Review by dlnr — COMMENTED** [review]
I really wonder if this wouldn't be a better fit in aip storybook, but at least now UX can test this component. How will you deploy this stuff this week for use in the form?

File: package.json [hunk 2]
```diff
@@ -111,21 +111,23 @@
...
 		"@amsterdam/asc-ui": "^0.36.0",
-		"@mui/icons-material": "^5.8.4",
-		"@mui/material": "^5.8.7",
-		"@mui/styled-engine-sc": "^5.8.0",
-		"@mui/x-data-grid": "^5.12.3",
+		"@mui/icons-material": "^5.10.3",
+		"@mui/material": "^5.10.5",
+		"@mui/styled-engine-sc": "^5.10.3",
+		"@mui/x-data-grid": "^5.17.3",
+		"asc-assets-beta": "npm:@amsterdam/asc-assets@^1.0.0-alpha.0",
```
[the remark this answers is not in the document]
Comment (reno1979) [reply package.json:118]: This repo has a select used by AIP, this new asyncSelect uses the same deps

File: package.json [hunk 3]
```diff
@@ -111,21 +111,23 @@
...
+		"asc-ui-beta": "npm:@amsterdam/asc-ui@^1.0.0-alpha.0",
 		"react": ">=17.0.2",
 		"react-dom": ">=17.0.2",
 		"react-dropzone": "^11.3.5",
 		"react-loading-skeleton": "^3.0.3",
 		"react-select": "5.2.2",
-		"styled-components": ">=5.2.3"
+		"styled-components": ">=5.3.5"
```
[the remark this answers is not in the document]
Comment (reno1979) [reply package.json:125]: There is no automatic style testing, I did not see any changes but again no actual style testing available so with each change we will just have to check manualy

File: src/form/AsyncSelect/AsyncSelect.tsx [hunk 4]
```diff
@@ -0,0 +1,43 @@
+import React from 'react';
+import { ComponentProps } from 'react';
+import { components, DropdownIndicatorProps } from 'react-select';
+import ReactSelectAsyncSelect from 'react-select/async';
+import { DropdownIndicatorStyle, getAsyncSelectStyle } from './AsyncSelectStyles';
+
+export type Props = {
+	error?: boolean;
+	zIndexMenu?: number;
```
[the remark this answers is not in the document]
Comment (reno1979) [reply src/form/AsyncSelect/AsyncSelect.tsx:9]: Use the other select component as base for this one, I am certain there was a valid reason for adding this

File: src/form/AsyncSelect/AsyncSelect.tsx (hunk 1 above)
Comment (reno1979) [reply src/form/AsyncSelect/AsyncSelect.tsx:18]: The todo's are not meant for this component, this is a draft and these todo's are more or less questions / what ifs..

File: package.json (hunk 3 above)
[the remark this answers is not in the document]
Comment (dlnr) [reply package.json:125]: I'm not suggesting we do any style testing, just asking if you manually checked the other components. Probably fine though.

File: src/form/AsyncSelect/AsyncSelect.stories.tsx [hunk 5]
```diff
@@ -0,0 +1,63 @@
+import React from 'react';
+import { ComponentStory, ComponentMeta } from '@storybook/react';
+import { AsyncSelect } from './AsyncSelect';
+
+type TypeaheadProps = {
+	_display: string;
```
Comment (fortunato) [inline src/form/AsyncSelect/AsyncSelect.stories.tsx:6]: What's the meaning of the prefixed underscore?

File: src/form/AsyncSelect/AsyncSelectStyles.ts [hunk 6]
```diff
@@ -0,0 +1,139 @@
+import { StylesConfig } from 'react-select';
+import { themeColor as themeColorAsc, themeSpacing as themeSpacingAsc, calculateFluidStyle } from 'asc-ui-beta';
+import styled from 'styled-components';
+import { Props } from './AsyncSelect';
+import { ascDefaultTheme } from 'asc-ui-beta';
+
+// Proxy to add the used theme
+function themeColor(type: string, variant?: string) {
```
Comment (fortunato) [inline src/form/AsyncSelect/AsyncSelectStyles.ts:8]: No need for @ts-ignore here. Please ask the team before resorting to these, thanks :pray: 

```typescript
// Proxy to add the used theme
function themeColor(type?: Parameters<typeof themeColorAsc>[0], variant?: Parameters<typeof themeColorAsc>[1]) {
	return themeColorAsc(type, variant)({ theme: ascDefaultTheme });
}
// Proxy to add the used theme
function themeSpacing(top: number, right?: number, bottom?: number, left?: number) {
	return themeSpacingAsc(top, right, bottom, left)({ theme: ascDefaultTheme });
}
```

This only leaves me with:
`TS2345: Argument of type '"level"' is not assignable to parameter of type 'ColorType | undefined'.`
From here:
```typescript
singleValue: (styles) => ({
	...styles,
	color: `${themeColor('level', 'level6')}`,
}),
```

File: tests/helpers/customRender.tsx [hunk 7]
```diff
@@ -3,7 +3,7 @@ import { render, RenderOptions, RenderResult } from '@testing-library/react';
 import { ThemeProvider } from '@amsterdam/asc-ui';
 import theme from '../../src/theme/theme';
 import { queries, Queries } from '@testing-library/dom';
-
+//@ts-ignore
```
Comment (fortunato) [inline tests/helpers/customRender.tsx:6]:
```typescript
const AllTheProviders = ({ children }: PropsWithChildren<{}>) => (
	<ThemeProvider overrides={theme}>{children}</ThemeProvider>
);
```

**Review by fortunato — CHANGES_REQUESTED** [review]
Looks good mostly :+1:

File: src/form/AsyncSelect/AsyncSelectStyles.ts (hunk 6 above)
Comment (fortunato) [reply src/form/AsyncSelect/AsyncSelectStyles.ts:8]: Do you why that error would show up?

File: src/form/AsyncSelect/AsyncSelect.stories.tsx (hunk 5 above)
Comment (reno1979) [reply src/form/AsyncSelect/AsyncSelect.stories.tsx:6]: I will remove this and replace it withj a __stubs__ , the data.amsterdam just returns data that way :(

File: src/form/AsyncSelect/AsyncSelectStyles.ts (hunk 6 above)
Comment (reno1979) [reply src/form/AsyncSelect/AsyncSelectStyles.ts:8]: you are asking me about TS, right :D

Comment (reno1979) [reply src/form/AsyncSelect/AsyncSelectStyles.ts:8]: fixed

**Review by fortunato — APPROVED** [review]

**Pull request merged**